units/patch_intent.go

```diff
@@ -488,73 +450,103 @@ func (j *patchIntentProcessor) finishPatch(ctx context.Context, patchDoc *patch.
 	return catcher.Resolve()
 }
 
-func (j *patchIntentProcessor) getPreviousPatchDefinition(project *model.Project, failedOnly bool) ([]patch.VariantTasks, error) {
-	previousPatch, err := patch.FindOne(patch.MostRecentPatchByUserAndProject(j.user.Username(), project.Identifier))
-	if err != nil {
-		return nil, errors.Wrap(err, "error querying for most recent patch")
+func (j *patchIntentProcessor) buildTasksandVariants(patchDoc *patch.Patch, project *model.Project) error {
+	var previousPatchStatus string
+	var err error
+
+	failedOnly := j.intent.RepeatFailedTasksAndVariants()
+
+	if j.intent.ReusePreviousPatchDefinition() || failedOnly {
+		previousPatchStatus, err = j.setToPreviousPatchDefinition(patchDoc, project, failedOnly)
+		if err != nil {
+			return err
+		}
 	}
-	if previousPatch == nil {
-		return nil, errors.Errorf("no previous patch available")
+
+	// verify that all variants exists
+	for _, buildVariant := range patchDoc.BuildVariants {
+		if buildVariant == "all" || buildVariant == "" {
+			continue
+		}
+		bv := project.FindBuildVariant(buildVariant)
+		if bv == nil {
+			return errors.Errorf("no such buildvariant matching '%s'", buildVariant)
+		}
 	}
-	var res []patch.VariantTasks
-	if failedOnly && !(previousPatch.Status == evergreen.PatchFailed) {
-		return res, nil
+
+	for _, bv := range patchDoc.RegexBuildVariants {
+		_, err := regexp.Compile(bv)
+		if err != nil {
+			return errors.Wrapf(err, "compiling buildvariant regex '%s'", bv)
+		}
 	}
+	for _, t := range patchDoc.RegexTasks {
+		_, err := regexp.Compile(t)
+		if err != nil {
+			return errors.Wrapf(err, "compiling task regex '%s'", t)
+		}
+	}
+
+	// if the user only wants failed tasks but the previous patch has no failed tasks, there is nothing to build
+	skipForFailed := failedOnly && previousPatchStatus != evergreen.PatchFailed
+
+	if len(patchDoc.VariantsTasks) == 0 && !skipForFailed {
```

Comment:
Is it possible to define reuse with other things? (regex etc?) Because then skipping this for reuse-failed likely isn't worth it / could introduce more weirdness

Reply:
I may be misunderstanding your comment, but if someone wants to restart only failed tasks, then no matter what else they specify, if there are only passing tests, then we always want to run nothing.

Reply:
Gotcha, yeah ticketing makes sense then 👍